Apps/TimelineApp/feature/user/src/main/java/uk/ac/cam/cares/jps/user/SensorSettingFragment.java

```diff
@@ -225,46 +252,76 @@ private void onRecordButtonClicked(View view) {
         });
     }
 
+    private void startRecording(List<SensorType> selectedSensorTypes) {
+        if (sensorViewModel.getIsRecording().getValue() != null && sensorViewModel.getIsRecording().getValue()) {
+            // Stop recording and reset sensors
+            sensorViewModel.stopRecording();
+            for (SensorItem item : adapter.getSensorItems()) {
+                item.setToggled(false); // Reset toggles
+            }
+            adapter.setTogglesEnabled(true);
+            binding.toggleAllBtn.setEnabled(true);
+        } else {
+            // Start recording only the selected sensors
+            sensorViewModel.startRecording();
+            adapter.setTogglesEnabled(false);
+            binding.toggleAllBtn.setEnabled(false);
+        }
+    }
+
     private void initPermissions() {
         permissionsMap.put(Permission.PermissionType.LOCATION_FINE, new Permission(Permission.PermissionType.LOCATION_FINE, requestFineLocationPermissionLauncher));
         permissionsMap.put(Permission.PermissionType.AUDIO, new Permission(Permission.PermissionType.AUDIO, requestAudioPermissionLauncher));
         permissionsMap.put(Permission.PermissionType.NOTIFICATION, new Permission(Permission.PermissionType.NOTIFICATION, requestNotificationPermissionLauncher));
     }
 
 
-    private void checkFineLocationPermission(Permission permission) {
+    private void checkFineLocationPermission(Permission permission, Runnable onPermissionGranted) {
         if (ContextCompat.checkSelfPermission(requireContext(), permission.permissionString) == PackageManager.PERMISSION_GRANTED) {
             permission.isGranted = true;
+            onPermissionGranted.run();
         } else if (ActivityCompat.shouldShowRequestPermissionRationale(requireActivity(), permission.permissionString)) {
             new MaterialAlertDialogBuilder(requireContext())
                     .setMessage(permission.explanation)
                     .setPositiveButton(uk.ac.cam.cares.jps.ui.R.string.ok, (dialogInterface, i) -> {
                         // some permissions (eg. notification) may be auto granted in lower sdk version
                         if (permission.permissionString.isEmpty()) {
                             permission.isGranted = true;
+                            onPermissionGranted.run();
                             return;
                         }
+                        pendingPermissionCallback = onPermissionGranted;
```

Comment:
The idea of using a global runnable object here is a bit weird. If the action for all permissions are the same, should just call fragment member function directly in the `ActivityResultLauncher` instead of create an global object and set it to null after finishing using, which makes the workflow confusing and have the risk of null pointer exception